raiden_contracts/contracts/services/UserDeposit.sol

```diff
@@ -0,0 +1,172 @@
+    Token public token;
+
+    // Trusted contract (can execute `transfer`)
+    address public msc_address;
+
+    // Keep track of balances
+    mapping(address => uint256) public balances;
+    mapping(address => WithdrawPlan) public withdraw_plans;
+
+    /*
+     *  Structs
+     */
+    struct WithdrawPlan {
+        uint256 amount;
+        uint256 withdraw_block;  // earliest block at which withdraw is allowed
+    }
+
+    /*
+     *  Events
+     */
+
+    event BalanceReduced(address indexed owner, uint newBalance);
+    event WithdrawPlanned(address indexed withdrawer, uint plannedBalance);
+
+    /*
+     *  Modifiers
+     */
+
+    modifier canTransfer() {
+        require(msg.sender == msc_address);
+        _;
+    }
+
+    /*
+     *  Constructor
+     */
+
+    /// @notice Set the default values for the smart contract
+    /// @param _token_address The address of the token to use for rewards
+    constructor(address _token_address)
+        public
+    {
+        // check token contract
+        require(_token_address != address(0x0));
+        require(contractExists(_token_address));
+        token = Token(_token_address);
+        require(token.totalSupply() > 0); // Check if the contract is indeed a token contract
+    }
+
+    /// @notice Specify trusted contracts. This has to be done outside of the
+    /// constructor to avoid cyclic dependencies.
+    /// @param _msc_address Address of the MonitoringService contract
+    function init(address _msc_address)
+        external
+    {
+        // prevent changes of trusted contract after initialization
+        require(msc_address == address(0x0));
+
+        // check monitoring service contract
+        require(_msc_address != address(0x0));
+        require(contractExists(_msc_address));
+        msc_address = _msc_address;
+    }
+
+    /// @notice Deposit tokens. Idempotent function that sets the
+    /// total_deposit of tokens of the beneficiary.
+    /// Can be called by anyone several times and on behalf of other accounts
+    /// @param beneficiary The account benefiting from the deposit
+    /// @param total_deposit The amount tokens in the user's deposit after the call
+    function deposit(address beneficiary, uint256 total_deposit)
+        external
+    {
+        require(total_deposit > balances[beneficiary]);
+
+        // Calculate the actual amount of tokens that will be transferred
+        uint256 added_deposit = total_deposit - balances[beneficiary];
+
+        balances[beneficiary] += added_deposit;
+        require(token.transferFrom(msg.sender, address(this), added_deposit));
+    }
+
+    /// @notice Internally transfer deposits between two addresses.
+    /// @param sender Account from which the amount will be deducted
+    /// @param receiver Account to which the amount will be credited
+    /// @param amount Amount of tokens to be transferred
+    /// @return true if transfer has been done successfully, otherwise false
```

Comment:
When I read this, I might believe all failures return false.  Document the throwing behavior as well.

Reply:
I'm not seeing any comments for the possible `require()` failure.  In the end people's fund might be locked up because of this.